### PR TITLE
[ADT] Fix ArrayRef<T>::slice

### PR DESCRIPTION
**Pull request by FLZ101**

### CODE REVIEW

File: llvm/include/llvm/ADT/ArrayRef.h [hunk 1]
```diff
@@ -198,7 +198,10 @@ namespace llvm {
     }
 
     /// slice(n) - Chop off the first N elements of the array.
-    ArrayRef<T> slice(size_t N) const { return slice(N, size() - N); }
+    ArrayRef<T> slice(size_t N) const {
+      assert(N <= size() && "Invalid specifier");
+      return slice(N, size() - N);
```
Comment (kuhar) [inline llvm/include/llvm/ADT/ArrayRef.h:203]: Instead, we could call into `drop_front` which does the same thing. Or make `drop_front` call slice.

**Review by kuhar — CHANGES_REQUESTED** [review]
Could you update the PR description and explain what the issue is?

File: llvm/include/llvm/ADT/ArrayRef.h (hunk 1 above)
Comment (kuhar) [reply llvm/include/llvm/ADT/ArrayRef.h:203]: I think `drop_front` is more idiomatic and we could probably mark the single-argument slice as deprecated and eventually remove it. Although that's all out of scope for this PR. cc: @kazutakahirata

**Comment by FLZ101** [comment]
> Could you update the PR description and explain what the issue is?

`slice(N, size() - N)` will never fail the assertion `assert(N+M <= size() && "Invalid specifier")` above, even `N > size()`.

File: llvm/include/llvm/ADT/ArrayRef.h (hunk 1 above)
Comment (FLZ101) [reply llvm/include/llvm/ADT/ArrayRef.h:203]: Agree.

**Review by kuhar — APPROVED** [review]
Looks good but please add the explanation from the comment to the commit description when merging.

**Comment by FLZ101** [comment]
> Looks good but please add the explanation from the comment to the commit description when merging.

Done.

**Comment by FLZ101** [comment]
@kuhar Could you merge this PR?

**Pull request merged**